networkx/algorithms/assortativity/correlation.py

```diff
@@ -254,20 +262,21 @@ def attribute_ac(M):
     return r
 
 
-def numeric_ac(M):
+def numeric_ac(M, mapping):
     # M is a numpy matrix or array
     # numeric assortativity coefficient, pearsonr
     import numpy as np
 
     if M.sum() != 1.0:
         M = M / float(M.sum())
     nx, ny = M.shape  # nx=ny
-    x = np.arange(nx)
-    y = np.arange(ny)
+    x = np.array(list(mapping.keys()))
+    y = x  # x and y have the same support
+    idx = list(mapping.values())
     a = M.sum(axis=0)
     b = M.sum(axis=1)
-    vara = (a * x ** 2).sum() - ((a * x).sum()) ** 2
-    varb = (b * x ** 2).sum() - ((b * x).sum()) ** 2
+    vara = (a[idx] * x ** 2).sum() - ((a[idx] * x).sum()) ** 2
+    varb = (b[idx] * x ** 2).sum() - ((b[idx] * x).sum()) ** 2
```

Comment:
I know the values are the same, but could you change “x” to “y” in “varb”?
This also makes me wonder if the mixing matrix for a directed graph ought to have in_degree for rows and out_degree for columns.  But I think that is beyond this issue.  It might be a topic of a whole paper on mixing over graphs (if it hasn’t been written already).

Reply:
Ok, I will change it.
In the current implementation, a mixing matrix is squared and supports (sets of possible values) for in_degree and out_degree are the same. But there is another way to represent a mixing matrix in a non-squared view. That is as you said, in_degree is for rows and out_degree is for columns (or vice versa). Both are correct, and do not affect assortativity coefficient. The disadvantage of a non-squared view is that it requires separate mappings for rows and columns, say two variables `out_mapping` and `in_mapping` instead of a single `mapping`. The advantage of a non-squared view is that a matrix is much smaller in some limit cases, say in a graph with many trees. For example, a graph

<img src='https://user-images.githubusercontent.com/12996098/123511014-80841180-d687-11eb-92d7-4b8de61bad3f.png' width=400>

has a 3x3 degree mixing matrix 
```
mapping={1: 0, 3: 1, 4: 2}
```
```
[[0.   0.   0.  ]
 [0.43 0.   0.  ]
 [0.57 0.   0.  ]]
```
or a 2x1 degree mixing matrix
```
out_mapping = {3: 0, 4: 1}
in_mapping = {1: 0}
```
```
[[0.43]
 [0.57]]
```
As we see, if we increase the number of such trees with different number of leafs, the non-squared mixing matrix will still be just a one-column matrix.

Reply:
Nice description -- Thanks! 

I think this is ready to merge.  Anything else you know of?

Reply:
Yes, it is ready to merge